### PR TITLE
WIP: Refactor/result manager

### PR DESCRIPTION
**Pull request by castellinilinguini**
# Description

Pull request to merge Refactor/result_manager into main. This

* improves clarity and brevity in the folder structure and file names created by EPI, while simultaneously reducing unnecessary redundancies in the data written to file.
* splits up the overly long and complicated ResultManager class into two new classes, OutputWriter and ResultReader, thus improving separation of concerns.
* removes the support for lists of slices, a feature that was hardly useable because there was no proper interface to pass other important inference parameters separately for each slice. This shortens the code significantly and also makes the EPI interface a lot easier.
* improves some of the tests qualitatively.

Fixes # (issue)

Prepares fixing #146.

## Type of change

Please tick at least one of the options. Delete the other options.

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

All unit tests pass. Some existing tests have been extended to cover the newly implemented changes.

## Checklist For Contributor

- [ ] My code follows the style guidelines of this project (I installed pre-commit)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I wrote my changes in the changelog in the `[unreleased]` section
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes

## Checklist For Maintainers

- [ ] Continuous Integration (CI) is successfully running
- [ ] Do we want to release/tag a new version? [✔️ / ❌]
  - [ ] If yes, add a release to the changelog and set the new version in pyproject.toml. After the merge, tag the release!

### CODE REVIEW

**Comment by castellinilinguini** [comment]
@kaiserls Would you mind checking if the implemented changes look ok?

**Comment by castellinilinguini** [comment]
Aaaaaand apparently my pre-commit installation is broken, I'll look into this tomorrow.

**Review by kaiserls — CHANGES_REQUESTED** [review]
Hi @castellinilinguini ,
Thank you for your work on the ResultManager refactoring and for reducing the overall code complexity. Splitting the logic into OutputWriter and ResultReader and removing the complex multi-slice support has significantly improved the code's accessibility and maintainability.

Here are a few points we should discuss during our meeting tomorrow:
- How to couple the functions in result_manager_utils.py with the result managers again? They are implementation details and should be easily exchangeable and customizable.
- There are some newly added result files in the tutorial_material. Are they needed for the changes introduced in this pull request?
- The renaming from "simulation results" to"pushforward" leads to changes in files not really relevant to this pull request like base_model.py and kde.py. It depends on how we merge this pull request and #143, whether we should shift these changes to another merge request.

For details please take a look at the comments belonging to the review.

One minor detail:
- Would you support a symmetric naming for the two new classes, e.g. ResultWriter and ResultReader?
Thank you again for your efforts on this! I’m looking forward to our discussion tomorrow.
Greetings

File: eulerpi/core/kde.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 """This module provides functions to handle the Kernel Densitiy Estimation (KDE_) in EPI.
 
-    It is used in the EPI algorithm to :py:func:`eulerpi.core.transformations.evaluate_density <evaluate the density>` of the transformed data distribution at the simulation results.
+    It is used in the EPI algorithm to :py:func:`eulerpi.core.transformations.evaluate_density <evaluate the density>` of the transformed data distribution at the pushforward.
```
[the remark this answers is not in the document]
Comment (castellinilinguini) [reply eulerpi/core/kde.py:3]: @shoepfl and I agreed that from a user perspective, “simulation result” is not a particularly clear descriptor, especially for the files (could just as well be the result of the inference, etc.). On top, the models are called "model" and not simulation, so it's unclear what simulation actually refers to. In context of EPI, we believe pushforward is the more accurate label.

File: eulerpi/core/result_managers/result_manager_utils.py
[the remark this answers is not in the document]
Comment (castellinilinguini) [reply eulerpi/core/result_managers/result_manager_utils.py:1]: I agree! Let's discuss how we could achieve this in our call :)

File: eulerpi/core/sampling.py [hunk 2]
```diff
@@ -17,14 +17,15 @@
 from os import path
 
 import emcee
+from eulerpi.core.result_managers.result_manager_utils import get_run_path
```
[the remark this answers is not in the document]
Comment (castellinilinguini) [reply eulerpi/core/sampling.py:20]: Yep, I think we should probably add a function to get the run path to the output writer to decouple this.

File: docs/source/tutorial_material/Output/Temperature/default_run_slice_Q0/DensityEvals/raw_density_evals_0.csv
[the remark this answers is not in the document]
Comment (castellinilinguini) [reply docs/source/tutorial_material/Output/Temperature/default_run_slice_Q0/DensityEvals/raw_density_evals_0.csv:1]: Ah, yes, the old files should probably be deleted, this is due to the change in the naming system.

File: eulerpi/core/result_managers/result_manager_utils.py
[the remark this answers is not in the document]
Comment (kaiserls) [reply eulerpi/core/result_managers/result_manager_utils.py:1]: Discussed Solution: Put the functions into a class e.g. `PathManager` and pass it to the Reader and Writer constructor -> Customization possible and Default behavior is sensible.

**Comment by castellinilinguini** [comment]
We employ a different merging strategy now: instead of merging this branch into main, the changes will be incorporated into the refactor/inference branch and then later merged into main.

**Pull request closed without merging**